### PR TITLE
Standardize support for record override

### PR DESCRIPTION
**Pull request by Gabriel439**
Fixes https://github.com/dhall-lang/dhall-lang/issues/382

This adds a new record instantiation operator of the form
`T::r` which is syntactic sugar for `(T.default ⫽ r) : T.Type)`

The motivation for this change is to handle configuration formats
with a large number of defaultable fields (especially if the field
is `Optional` and the default is `None U`) so that users can omit any
all default-valued fields when creating the record.

```
let Example =
      { Type = { foo : Text, bar : Optional Natural }
      , default = { bar = None Natural }
      }

    -- No need to specify `bar` since it will default to `None Natural`
in  Example::{ foo = "ABC" }
```

This syntax more closely resembles the common JSON/YAML idiom of
omitting default-valued fields.

### CODE REVIEW

**Comment by Gabriel439** [comment]
Matching change to the Haskell implementation: https://github.com/dhall-lang/dhall-haskell/pull/1375

File: standard/README.md [hunk 1]
```diff
@@ -117,6 +117,7 @@ a, b, f, l, r, e, t, u, A, B, E, T, U, c, i, o
   / t.x                               ; Field selection
   / t.{ xs… }                         ; Field projection
   / t.(s)                             ; Field projection by type
+  / T⸬r                               ; Record override
```
Comment (sjakobi) [inline standard/README.md]: Is this the right operator? I thought this was about `::` or, in unicode, `∷`.

The same operator appears in a few more places too.

File: tests/normalization/success/simple/overrideA.dhall [hunk 2]
```diff
@@ -0,0 +1,10 @@
+let Example =
+      { Type = { name : Text, id : Optional Natural }
+      , default = { name = "", id = None Natural }
+      }
+
+in  { example0 = Example∷{=}
```
Comment (sjakobi) [inline tests/normalization/success/simple/overrideA.dhall]: This operator doesn't actually appear in the standard, although I do like it better than the triple-dot (`⸬`), which I find very hard to read and rather different than the ASCII version `::`.

IMHO `::` is already visually pleasing enough that I don't really see the need for a unicode symbol version.

File: standard/shift.md [hunk 3]
```diff
@@ -328,6 +328,11 @@ The remaining rules are:
     ↑(d, x, m, t₀.{ xs… }) = t₁.{ xs… }
 
 
+    ↑(d, x, m, t₀) = t₁
+    ───────────────────────
+    ↑(d, x, m, t₀⸬x) = t₁⸬x
```
Comment (sjakobi) [inline standard/shift.md]: Shouldn't `x` be shifted too? It's a bit confusing that `x` is both the variable and the operand BTW (same situation above in the field access rule).

File: standard/substitution.md [hunk 4]
```diff
@@ -309,6 +309,11 @@ The remaining rules are:
     (t₀.{ xs… })[x@n ≔ e] = t₁.{ xs… }
 
 
+    t₀[x@n ≔ e] = t₁
+    ────────────────────────
+    (t₀⸬x₀)[x@n ≔ e] = t₁⸬x₀
```
Comment (sjakobi) [inline standard/substitution.md]: Shouldn't `x₀` get a substitution too?

**Comment by Gabriel439** [comment]
@sjakobi: How about "record completion operator"?

**Comment by sjakobi** [comment]
> How about "record completion operator"?

That's better IMO! :+1: 

---

I also think there probably should be a check that the `default`'s field types match the `Type`. Otherwise I really expect that some user is going to run into this gotcha and wonder why Dhall didn't catch it.

**Comment by Nadrieril** [comment]
The semantics of this operator are quite unique. It looks like the `Example` record is considered as a sort of ML-style module ?
I'm weakly opposed to this very ad-hoc looking operator, but since we don't have a strong stance on backward-compatibility we can always come up with a more general solution later.

**Comment by Gabriel439** [comment]
@Nadrieril: It is an ML-style module in the sense that it can store both types and terms.  However, I think the more direct inspiration is the NixOS module system, where you declare options like this:

```nix
{ lib, ... }:

{ options = {
    foo = lib.mkOption {
      type = lib.types.str;

      default = "bar";

      description = ''
        Description of the option
      '';
    };
  };
}
```

The names for the fields are directly inspired by the names for the NixOS option fields (and you can imagine us potentially adding support for a `description` field later for machine-readable documentation, just like NixOS options).

**Comment by Gabriel439** [comment]
@sjakobi: I think if we add that check, the check should be added as part of the underlying `//` operator (in a separate proposal).  In other words, we could update `//` to require that when overriding a field you don't change its type.  If you did that then an ill-typed `default` field would never type-check, whether or not the default were overridden.

I would prefer to keep the current operator to be syntactic sugar, instead of adding custom type-checking rules for the operator.  Part of the value of this proposal is that it's very simple to implement and it's also easy for users to reason about because it codifies an existing idiom that they have likely already been using.

**Comment by sjakobi** [comment]
> we could update `//` to require that when overriding a field you don't change its type

That might be a good idea, but it seems like quite the breaking change.

> I would prefer to keep the current operator to be syntactic sugar, instead of adding custom type-checking rules for the operator. Part of the value of this proposal is that it's very simple to implement and it's also easy for users to reason about because it codifies an existing idiom that they have likely already been using.

I don't think that the additional type-check would really affect the mental model. It's just an additional safeguard to rule out buggy `Type`-`default`-constructions.

It would be easy to add this safeguard at a later point though, so I guess it's fine to leave it off for now.

**Comment by Nadrieril** [comment]
@Gabriel439 Ok, makes sense. Could we then similarly require the use of a `mkOption` (we might call it something like `mkDefaultableRecord`) function before the `::` operation is allowed ? This function would ensure that the provided record only has those two fields, and would help discoverability. It might even be able to check that the default record has the correct type ? I'm not sure about that one.

**Comment by sjakobi** [comment]
> Could we then similarly require the use of a `mkOption` (we might call it something like `mkDefaultableRecord`) function before the `::` operation is allowed ?

What would that function do?

> This function would ensure that the provided record only has those two fields

Why would you want the left operand to be restricted to the `Type` and `default` fields?

> It might even be able to check that the default record has the correct type ?

That would of course be useful IMHO.

**Comment by Nadrieril** [comment]
> Why would you want the left operand to be restricted to the Type and default fields?
Because otherwise it feels ugly to me to duck-type the record like that and pick out a default field, hoping that it works as expected.

But of that doesn't seem worthwhile, we could still add a prelude function `makeDefaultRecord = \(T: Type) -> \(x: T) -> { Type = T, default = x }`, and recommend its use when using the `::` operator. This would at least typecheck the default record.

**Comment by sjakobi** [comment]
> But of that doesn't seem worthwhile, we could still add a prelude function `makeDefaultRecord = \(T: Type) -> \(x: T) -> { Type = T, default = x }`, and recommend its use when using the `::` operator. This would at least typecheck the default record.

That wouldn't type-check in the usual case where the `default` contains only some of the `Type`'s fields.

**Comment by Nadrieril** [comment]
Oh, I didn't realize we wanted to allow that. Then nevermind

**Comment by Gabriel439** [comment]
@sjakobi: I actually think it would be fine to require (in a separate proposal) that `//` doesn't change the type of a field, because:

* I believe the vast majority of users don't use `//` to change a field's type

  In fact, I believe most users would prefer Dhall to warn them if they mistakenly changed the field's type

* This has already been requested as part of https://github.com/dhall-lang/dhall-lang/issues/436

  Note that the above proposal requested two separate things: checking that the field already existed and it had the same type.  We'd only be implementing the latter check since requiring that the field exist would disrupt far more existing code in the wild.

* Users can still override a field's type if they *really* want to, by projecting out all fields except the one they want to override

File: standard/type-inference.md [hunk 5]
```diff
@@ -583,6 +583,29 @@ of the record must have the same type, which in turn must be a `Type`.
     Γ ⊢ ( toMap e : T₁ ) : T₁
 
 
+You can complete a record literal using the record completion operator (`T::r`),
+which is syntactic sugar for `(T.default ⫽ r) : T.Type`.  The motivation forx
```
Comment (sjakobi) [inline standard/type-inference.md]:
```suggestion
which is syntactic sugar for `(T.default ⫽ r) : T.Type`.  The motivation for
```

**Comment by PierreR** [comment]
@Gabriel439 Just want to say I believe this change is wonderful ...

It is just syntactic sugar but I believe it is going to allow me to hide under 'Example' a bit of details I don't want normal user (the one that doesn't know anything about Dhall) to see.
Thanks !

**Comment by Gabriel439** [comment]
@PierreR: You're welcome!  I plan to merge this as soon after fixing the name of the `Override` constructor to be `RecordCompletion`

**Pull request merged**